pkg/nsx/services/inventory/inventory.go

```diff
@@ -0,0 +1,460 @@
+		if err != nil {
+			log.Error(err, "Failed to split meta namespace key", "key", key)
+			continue
+		}
+		switch key.InventoryType {
+
+		case ContainerApplicationInstance:
+			pod := &corev1.Pod{}
+			err := s.Client.Get(context.TODO(), types.NamespacedName{Name: name, Namespace: namespace}, pod)
+			if apierrors.IsNotFound(err) ||
+				((err == nil) && (string(pod.UID) != externalId)) {
+				s.DeleteResource(externalId, ContainerApplicationInstance)
+			} else if err == nil {
+				retry := s.AddPod(pod)
+				if retry {
+					retryKeys.Insert(key)
+				}
+			} else {
+				log.Error(err, "Unexpected error is found while processing pod")
+			}
+
+		}
+	}
+
+	err := s.sendNSXRequestAndUpdateInventoryStore()
+	if err != nil {
+		return bufferedKeys, err
+	}
+
+	return retryKeys, err
+}
+
+func (s *InventoryService) DeleteResource(external_id string, resource_type InventoryType) error {
+	log.V(1).Info("Delete inventory resource", "resource_type", resource_type, "external_id", external_id)
+	var inventoryObject interface{} = nil
+	exists := false
+	switch resource_type {
+
+	case ContainerApplicationInstance:
+		inventoryObject = s.applicationInstanceStore.GetByKey(external_id)
+		if inventoryObject != nil {
+			exists = true
+		}
+	default:
+		return fmt.Errorf("unknown resource_type : %v for external_id %s", resource_type, external_id)
+	}
+
+	if exists {
+		deletedInfo := make(map[string]interface{})
+		deletedInfo["resource_type"] = resource_type
+		deletedInfo["external_id"] = external_id
+		s.requestBuffer = append(s.requestBuffer, containerinventory.ContainerInventoryObject{ContainerObject: deletedInfo, ObjectUpdateType: operation_delete})
+		s.pending_delete[external_id] = inventoryObject
+
+		// Update Pods which used to be connected to this removed service.
+		if resource_type == ContainerApplication {
+			namespaceId := inventoryObject.(containerinventory.ContainerApplication).ContainerProjectId
+			if namespaceId != "" {
+				/*
+					project, exists, _ := s.projectStore.GetByKey(namespaceId)
+					if exists {
+						s.removeServiceIdForPods(external_id, namespaceId, project.(containerinventory.ContainerProject).DisplayName, []string{})
+					}
+				*/
+			} else {
+				return fmt.Errorf("cannot update Pods for removed service id : %s, name : %s because namespaceId is empty", external_id, inventoryObject.(containerinventory.ContainerApplication).DisplayName)
+			}
+		}
+	}
+	return nil
+}
+
+func (s *InventoryService) sendNSXRequestAndUpdateInventoryStore() error {
+	if len(s.requestBuffer) > 0 {
+		log.Info("Send update to NSX clusterId ", "ContainerInventoryData", s.requestBuffer)
+		// TODO, check the context.TODO() be replaced by InventoryClient related todo
+		resp, err := s.NSXClient.InventoryClient.ContainerInventoryApi.AddContainerInventoryUpdateUpdates(context.TODO(), s.NSXConfig.Cluster, containerinventory.ContainerInventoryData{ContainerInventoryObjects: s.requestBuffer})
+
```

Comment:
I wonder AddContainerInventoryUpdateUpdates if the batch operation request is asynchronous，
1. How do you know if the process is successful or not?
2. If failed, after updateInventoryStore, how do you guarantee the consistency of the local store and NSX side?

Reply:
it should be sync.
The error indicated if the process status.
Only error is nil, the data will be saved to store
else will be retried.